language/vm/src/views.rs

```diff
@@ -645,32 +653,78 @@ impl<'a, T: ModuleAccess> SignatureTokenView<'a, T> {
     pub fn resolve_in_module(&self, other_module: &T) -> Option<SignatureToken> {
-        if let Some(struct_handle) = self.struct_handle() {
-            // Token contains a struct handle from `self.module`. Need to resolve inside
-            // `other_module`. We do this by normalizing the struct in `self.token`, then
-            // searching for the normalized representation inside `other_module`
-            // TODO: do we need to resolve `self.token`'s type actuals?
-            let type_actuals = Vec::new();
-            ModuleView::new(other_module)
-                .resolve_struct(&struct_handle.normalize_struct())
-                .map(|handle_idx| SignatureToken::Struct(handle_idx, type_actuals))
-        } else {
-            Some(self.token.clone())
+        match self.signature_token() {
+            SignatureToken::Struct(sh_idx, type_params) => {
+                // Token contains a struct handle from `self.module`. Need to resolve inside
+                // `other_module`. We do this by normalizing the struct in `self.token`, then
+                // searching for the normalized representation inside `other_module`
+                let type_actuals = type_params
+                    .iter()
+                    .map(|sig_token| {
+                        Self::new(other_module, sig_token).resolve_in_module(other_module)
+                    })
+                    .collect::<Option<Vec<_>>>()?;
+                ModuleView::new(other_module)
+                    .resolve_struct(
+                        &StructHandleView::from_handle_idx(self.module, *sh_idx)
+                            .normalize_struct(Vec::new()),
+                    )
+                    .map(|handle_idx| SignatureToken::Struct(handle_idx, type_actuals))
+            }
+            _ => None,
         }
     }
+
+    /// Return a type tag corresponding to the type of `self` instantiated with the type actuals
+    /// in `type_actuals_tags`.
+    /// Returns None if `self` is not a struct type or if the size of `type_actual_tags` does not
+    /// match the number of type actuals that the struct expects.
+    pub fn derive_type_tag(&self, type_actual_tags: &[TypeTag]) -> Option<TypeTag> {
+        use SignatureToken::*;
+        Some(match self.token {
+            Bool => TypeTag::Bool,
+            Address => TypeTag::Address,
+            U8 => TypeTag::U8,
+            U64 => TypeTag::U64,
+            U128 => TypeTag::U128,
+            ByteArray => TypeTag::ByteArray,
+            TypeParameter(idx) => return type_actual_tags.get(*idx as usize).cloned(),
+            Reference(_) | MutableReference(_) =>
+            // don't derive a tag for references, only structs
+            {
+                return None
+            }
+            Struct(idx, type_actual_tokens) => {
+                if type_actual_tokens.len() != type_actual_tags.len() {
+                    // wrong number of type actuals
+                    return None;
```

Comment:
assert instead? This seems to me a programmer error.

Reply:
Assert is more stringent than the old semantics (which returned `None`), but I will try it and see if it breaks anything.

Reply:
Interestingly, a lot of tests fail if you assert instead of returning `None` here! Definitely worth some investigation in the future.